File: src/package_tools.py
"""
Package Tools - инструменты для управления пакетами.
Поддержка: pip (Python), npm (Node.js), apt (Debian/Ubuntu).
"""
from __future__ import annotations

import json
import subprocess
import sys
from dataclasses import dataclass
from pathlib import Path


@dataclass
class ToolResult:
    """Результат выполнения инструмента."""
    success: bool
    output: str
    error: str = ""


class PackageInstallTool:
    """Установка пакетов через различные менеджеры."""

    def __init__(self, workspace_root: str | None = None):
        self.workspace_root = Path(workspace_root).resolve() if workspace_root else Path.cwd()

    def install(
        self,
        package: str,
        manager: str = "auto",
        dev: bool = False,
        global_install: bool = False
    ) -> ToolResult:
        """Установить пакет.

        Args:
            package: Имя пакета для установки
            manager: Менеджер пакетов (auto, pip, npm, apt)
            dev: Установить как dev-зависимость (для npm)
            global_install: Глобальная установка

        Returns:
            ToolResult с результатом установки
        """
        # Автоопределение менеджера
        if manager == "auto":
            manager = self._detect_package_manager()

        if manager == "pip":
            return self._install_pip(package, global_install)
        elif manager == "npm":
            return self._install_npm(package, dev, global_install)
        elif manager == "apt":
            return self._install_apt(package)
        else:
            return ToolResult(
                success=False,
                output="",
                error=f"Unknown package manager: {manager}. Supported: pip, npm, apt"
            )

    def _detect_package_manager(self) -> str:
        """Автоматически определить менеджер пакетов."""
        # Проверяем Python проекты
        if (self.workspace_root / "requirements.txt").exists() or \
           (self.workspace_root / "pyproject.toml").exists() or \
           (self.workspace_root / "setup.py").exists():
            return "pip"

        # Проверяем Node.js проекты
        if (self.workspace_root / "package.json").exists():
            return "npm"

        # По умолчанию pip (т.к. мы в Python окружении)
        return "pip"

    def _install_pip(self, package: str, global_install: bool) -> ToolResult:
        """Установить Python пакет через pip."""
        try:
            cmd = [sys.executable, "-m", "pip", "install"]

            # Если не глобальная установка - используем --user
            if not global_install:
                cmd.append("--user")

            cmd.append(package)

            result = subprocess.run(
                cmd,
                cwd=str(self.workspace_root),
                capture_output=True,
                text=True,
                timeout=300  # 5 минут таймаут
            )

            success = result.returncode == 0

            output = result.stdout
            if result.stderr:
                output += f"\n{result.stderr}"

            return ToolResult(
                success=success,
                output=output,
                error="" if success else f"Failed to install {package}"
            )

        except FileNotFoundError:
            return ToolResult(
                success=False,
                output="",
                error="pip not found. Install Python first."
            )
        except subprocess.TimeoutExpired:
            return ToolResult(
                success=False,
                output="",
                error="Installation timeout (5 minutes)"
            )
        except Exception as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Error installing package: {str(e)}"
            )

    def _install_npm(self, package: str, dev: bool, global_install: bool) -> ToolResult:
        """Установить Node.js пакет через npm."""
        try:
            cmd = ["npm", "install"]

            if global_install:
                cmd.append("-g")
            elif dev:
                cmd.append("--save-dev")

            cmd.append(package)

            result = subprocess.run(
                cmd,
                cwd=str(self.workspace_root),
                capture_output=True,
                text=True,
                timeout=300
            )

            success = result.returncode == 0

            output = result.stdout
            if result.stderr:
                output += f"\n{result.stderr}"

            return ToolResult(
                success=success,
                output=output,
                error="" if success else f"Failed to install {package}"
            )

        except FileNotFoundError:
            return ToolResult(
                success=False,
                output="",
                error="npm not found. Install Node.js first."
            )
        except subprocess.TimeoutExpired:
            return ToolResult(
                success=False,
                output="",
                error="Installation timeout (5 minutes)"
            )
        except Exception as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Error installing package: {str(e)}"
            )

    def _install_apt(self, package: str) -> ToolResult:
        """Установить системный пакет через apt."""
        try:
            # apt требует sudo
            cmd = ["sudo", "apt-get", "install", "-y", package]

            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=600  # 10 минут для системных пакетов
            )

            success = result.returncode == 0

            output = result.stdout
            if result.stderr:
                output += f"\n{result.stderr}"

            return ToolResult(
                success=success,
                output=output,
                error="" if success else f"Failed to install {package}"
            )

        except FileNotFoundError:
            return ToolResult(
                success=False,
                output="",
                error="apt-get not found. Only works on Debian/Ubuntu."
            )
        except subprocess.TimeoutExpired:
            return ToolResult(
                success=False,
                output="",
                error="Installation timeout (10 minutes)"
            )
        except Exception as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Error installing package: {str(e)}"
            )


class PackageSearchTool:
    """Поиск пакетов в репозиториях."""

    def search(self, query: str, manager: str = "pip") -> ToolResult:
        """Поиск пакетов.

        Args:
            query: Поисковый запрос
            manager: Менеджер пакетов (pip, npm)

        Returns:
            ToolResult со списком найденных пакетов
        """
        if manager == "pip":
            return self._search_pip(query)
        elif manager == "npm":
            return self._search_npm(query)
        else:
            return ToolResult(
                success=False,
                output="",
                error=f"Search not supported for {manager}"
            )

    def _search_pip(self, query: str) -> ToolResult:
        """Поиск Python пакетов."""
        try:
            # pip search был отключён, используем PyPI API
            import requests
            response = requests.get(
                f"https://pypi.org/pypi/{query}/json",
                timeout=10
            )

            if response.status_code == 200:
                data = response.json()
                info = data.get("info", {})
                output = f"Package: {info.get('name')}\n"
                output += f"Version: {info.get('version')}\n"
                output += f"Summary: {info.get('summary')}\n"
                output += f"Author: {info.get('author')}\n"
                output += f"License: {info.get('license')}\n"
                output += f"URL: {info.get('package_url')}\n"

                return ToolResult(success=True, output=output)
            else:
                return ToolResult(
                    success=False,
                    output="",
                    error=f"Package '{query}' not found on PyPI"
                )

        except Exception as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Error searching package: {str(e)}"
            )

    def _search_npm(self, query: str) -> ToolResult:
        """Поиск Node.js пакетов."""
        try:
            cmd = ["npm", "search", query, "--json"]

            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=30
            )

            if result.returncode == 0:
                packages = json.loads(result.stdout)
                if packages:
                    output = ""
                    for pkg in packages[:5]:  # Первые 5 результатов
                        output += f"Package: {pkg.get('name')}\n"
                        output += f"Version: {pkg.get('version')}\n"
                        output += f"Description: {pkg.get('description')}\n"
                        output += f"Author: {pkg.get('author', {}).get('name', 'N/A')}\n"
                        output += "\n"

                    return ToolResult(success=True, output=output)
                else:
                    return ToolResult(
                        success=False,
                        output="",
                        error=f"No packages found for '{query}'"
                    )
            else:
                return ToolResult(
                    success=False,
                    output="",
                    error=result.stderr
                )

        except Exception as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Error searching package: {str(e)}"
            )

File: src/package_tools.rs
//! Package Tools — установка/поиск пакетов через pip/npm/apt/cargo/go/yarn.
//! Слитый порт `package_tools.py` (install через pip/npm/apt + поиск) и
//! `advanced_tools.py::PackageInstallTool` (install через pip/npm/yarn/
//! cargo/go) — в оригинале это были два отдельных, местами дублирующих
//! друг друга класса `PackageInstallTool` в разных файлах с разным
//! набором менеджеров; здесь один инструмент со всеми менеджерами обоих.

use std::path::Path;
use std::time::Duration;
use tokio::process::Command;

use crate::tools::ToolResult;

fn detect_package_manager(workspace: &Path) -> &'static str {
    if workspace.join("requirements.txt").exists()
        || workspace.join("pyproject.toml").exists()
        || workspace.join("setup.py").exists()
    {
        "pip"
    } else if workspace.join("package.json").exists() {
        "npm"
    } else if workspace.join("Cargo.toml").exists() {
        "cargo"
    } else if workspace.join("go.mod").exists() {
        "go"
    } else {
        "pip"
    }
}

async fn run(cmd: &str, args: &[&str], cwd: &Path, timeout_secs: u64) -> ToolResult {
    let mut command = Command::new(cmd);
    command.args(args).current_dir(cwd);

    let output = match tokio::time::timeout(Duration::from_secs(timeout_secs), command.output()).await {
        Ok(Ok(o)) => o,
        Ok(Err(e)) if e.kind() == std::io::ErrorKind::NotFound => {
            return ToolResult::error(format!("`{}` не найден в PATH", cmd));
        }
        Ok(Err(e)) => return ToolResult::error(format!("Ошибка запуска `{}`: {}", cmd, e)),
        Err(_) => return ToolResult::error(format!("Таймаут выполнения `{}` ({}с)", cmd, timeout_secs)),
    };

    let mut out = String::from_utf8_lossy(&output.stdout).to_string();
    let stderr = String::from_utf8_lossy(&output.stderr);
    if !stderr.is_empty() {
        out.push('\n');
        out.push_str(&stderr);
    }

    if output.status.success() {
        ToolResult::success(out)
    } else {
        ToolResult { success: false, output: out, error: Some(format!("Команда завершилась с ошибкой (код {:?})", output.status.code())), images: Vec::new() }
    }
}

/// Установить пакет. `manager` = "auto" | "pip" | "npm" | "yarn" | "apt" | "cargo" | "go".
pub async fn package_install(
    workspace: &Path,
    package: &str,
    manager: &str,
    dev: bool,
    global_install: bool,
) -> ToolResult {
    let manager = if manager == "auto" { detect_package_manager(workspace) } else { manager };

    match manager {
        "pip" => {
            let mut args = vec!["-m", "pip", "install"];
            if !global_install {
                args.push("--user");
            }
            args.push(package);
            run("python3", &args, workspace, 300).await
        }
        "npm" => {
            let mut args = vec!["install"];
            if global_install {
                args.push("-g");
            } else if dev {
                args.push("--save-dev");
            }
            args.push(package);
            run("npm", &args, workspace, 300).await
        }
        "yarn" => {
            let mut args = vec!["add"];
            if dev {
                args.push("--dev");
            }
            args.push(package);
            run("yarn", &args, workspace, 300).await
        }
        "apt" => run("sudo", &["apt-get", "install", "-y", package], workspace, 600).await,
        "cargo" => run("cargo", &["add", package], workspace, 300).await,
        "go" => run("go", &["get", package], workspace, 300).await,
        other => ToolResult::error(format!(
            "Неизвестный менеджер пакетов: {}. Поддерживаются: pip, npm, yarn, apt, cargo, go",
            other
        )),
    }
}

/// Поиск пакета. `manager` = "pip" (через PyPI JSON API) | "npm" (через `npm search --json`).
pub async fn package_search(query: &str, manager: &str) -> ToolResult {
    match manager {
        "pip" => {
            let client = reqwest::Client::builder().timeout(Duration::from_secs(10)).build().unwrap_or_default();
            let url = format!("https://pypi.org/pypi/{}/json", query);
            let resp = match client.get(&url).send().await {
                Ok(r) => r,
                Err(e) => return ToolResult::error(format!("Ошибка запроса к PyPI: {}", e)),
            };
            if !resp.status().is_success() {
                return ToolResult::error(format!("Пакет '{}' не найден на PyPI", query));
            }
            let data: serde_json::Value = match resp.json().await {
                Ok(v) => v,
                Err(e) => return ToolResult::error(format!("Не удалось разобрать ответ PyPI: {}", e)),
            };
            let info = &data["info"];
            let output = format!(
                "Package: {}\nVersion: {}\nSummary: {}\nAuthor: {}\nLicense: {}\nURL: {}\n",
                info["name"].as_str().unwrap_or("?"),
                info["version"].as_str().unwrap_or("?"),
                info["summary"].as_str().unwrap_or("?"),
                info["author"].as_str().unwrap_or("?"),
                info["license"].as_str().unwrap_or("?"),
                info["package_url"].as_str().unwrap_or("?"),
            );
            ToolResult::success(output)
        }
        "npm" => {
            let result = run("npm", &["search", query, "--json"], Path::new("."), 30).await;
            if !result.success {
                return result;
            }
            let packages: Vec<serde_json::Value> = match serde_json::from_str(&result.output) {
                Ok(v) => v,
                Err(_) => return ToolResult::error("Не удалось разобрать вывод `npm search --json`"),
            };
            if packages.is_empty() {
                return ToolResult::error(format!("Пакеты по запросу '{}' не найдены", query));
            }
            let mut output = String::new();
            for pkg in packages.iter().take(5) {
                output.push_str(&format!(
                    "Package: {}\nVersion: {}\nDescription: {}\nAuthor: {}\n\n",
                    pkg["name"].as_str().unwrap_or("?"),
                    pkg["version"].as_str().unwrap_or("?"),
                    pkg["description"].as_str().unwrap_or("?"),
                    pkg["author"]["name"].as_str().unwrap_or("N/A"),
                ));
            }
            ToolResult::success(output)
        }
        other => ToolResult::error(format!("Поиск не поддерживается для менеджера: {}", other)),
    }
}
